File: services/backend/pkg/models/apps.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type LiveDemo struct {
	Label string `json:"label"`
	URL   string `json:"url"`
}

type AppLink struct {
	Label string `json:"label"`
	URL   string `json:"url"`
}

// DeploymentVariant defines a named deployment flavor for an app
// (e.g. "Standalone", "High Availability", "Mit SSO").
type DeploymentVariant struct {
	Name           string `json:"name"`
	Description    string `json:"description"`
	DockerCommand  string `json:"dockerCommand"`
	DockerNote     string `json:"dockerNote"`
	ComposeCommand string `json:"composeCommand"`
	ComposeNote    string `json:"composeNote"`
	HelmCommand    string `json:"helmCommand"`
	HelmNote       string `json:"helmNote"`
	HelmValues     string `json:"helmValues"`
}

// AppField holds a single key-value pair for the dynamic "Fachliche Details" section.
type AppField struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type Apps struct {
	bun.BaseModel `bun:"table:apps,alias:a"`

	ID              string     `bun:"id,pk" json:"id"`
	Name            string     `bun:"name,notnull" json:"name"`
	Description     string     `bun:"description" json:"description"`
	Categories      []string   `bun:"categories,array" json:"categories"`
	LiveUrl         string     `bun:"live_url" json:"liveUrl"`
	LiveDemos       []LiveDemo `bun:"live_demos,type:jsonb" json:"liveDemos"`
	RepoUrl         string     `bun:"repo_url" json:"repoUrl"`
	Repositories    []AppLink  `bun:"repositories,type:jsonb" json:"repositories"`
	CustomLinks     []AppLink  `bun:"custom_links,type:jsonb" json:"customLinks"`
	HelmRepo        string     `bun:"helm_repo" json:"helmRepo"`
	DockerRepo      string     `bun:"docker_repo" json:"dockerRepo"`
	DocsUrl         string     `bun:"docs_url" json:"docsUrl"`
	Icon            string     `bun:"icon" json:"icon"`
	TechStack       []string   `bun:"tech_stack,array" json:"techStack"`
	License         string     `bun:"license" json:"license"`
	MarkdownContent string     `bun:"markdown_content" json:"markdownContent"`
	// CustomFields stores all "Fachliche Details" as dynamic key-value pairs.
	// The field schema (labels, order) is managed via PlatformSettings.DetailFields.
	CustomFields           []AppField           `bun:"custom_fields,type:jsonb" json:"customFields"`
	Status                 string               `bun:"status" json:"status"`
	CustomDockerCommand    string               `bun:"custom_docker_command" json:"customDockerCommand"`
	CustomComposeCommand   string               `bun:"custom_compose_command" json:"customComposeCommand"`
	CustomHelmCommand      string               `bun:"custom_helm_command" json:"customHelmCommand"`
	CustomDockerNote       string               `bun:"custom_docker_note" json:"customDockerNote"`
	CustomComposeNote      string               `bun:"custom_compose_note" json:"customComposeNote"`
	CustomHelmNote         string               `bun:"custom_helm_note" json:"customHelmNote"`
	CustomHelmValues       string               `bun:"custom_helm_values" json:"customHelmValues"`
	HasDeploymentAssistant bool                 `bun:"has_deployment_assistant,notnull,default:true" json:"hasDeploymentAssistant"`
	ShowDocker             bool                 `bun:"show_docker,notnull,default:true" json:"showDocker"`
	ShowCompose            bool                 `bun:"show_compose,notnull,default:true" json:"showCompose"`
	ShowHelm               bool                 `bun:"show_helm,notnull,default:true" json:"showHelm"`
	Tags                   []string             `bun:"tags,array" json:"tags"`
	Collections            []string             `bun:"collections,array" json:"collections"`
	IsFeatured             bool                 `bun:"is_featured,notnull,default:false" json:"isFeatured"`
	RatingAvg              float64              `bun:"rating_avg,notnull,default:0" json:"ratingAvg"`
	RatingCount            int                  `bun:"rating_count,notnull,default:0" json:"ratingCount"`
	OwnerID                uuid.UUID            `bun:"owner_id,type:uuid,nullzero" json:"ownerId"`
	Owner                  *Users               `bun:"rel:belongs-to,join:owner_id=id" json:"owner"`
	IsLocked               bool                 `bun:"is_locked,notnull,default:false" json:"isLocked"`
	SkipLinkProbe          bool                 `bun:"skip_link_probe,notnull,default:false" json:"skipLinkProbe"`
	LinkProbeStatus        string               `bun:"link_probe_status,notnull,default:'unknown'" json:"linkProbeStatus"`
	KnownIssue             string               `bun:"known_issue" json:"knownIssue"`
	IsReuse                bool                 `bun:"is_reuse,notnull,default:false" json:"isReuse"`
	ReuseRequirements      string               `bun:"reuse_requirements" json:"reuseRequirements"`
	DeploymentVariants     []DeploymentVariant  `bun:"deployment_variants,type:jsonb" json:"deploymentVariants"`
	Version                string               `bun:"version,notnull,default:''" json:"version"`
	Changelog              string               `bun:"changelog,notnull,default:''" json:"changelog"`
	UpdatedAt              time.Time            `bun:"updated_at,nullzero,notnull,default:current_timestamp" json:"updatedAt"` // Virtual: populated by GetApp for the detail view
	RelatedApps            []AppRelationSummary `bun:"-" json:"relatedApps,omitempty"`
	AppGroups              []AppGroupSummary    `bun:"-" json:"appGroups,omitempty"`
	GitLabSync             *GitLabSyncSummary   `bun:"-" json:"gitLabSync,omitempty"`
	// Legacy detail-field aliases kept for JSON import/export compatibility.
	Focus           string `bun:"-" json:"focus"`
	AppType         string `bun:"-" json:"app_type"`
	UseCase         string `bun:"-" json:"use_case"`
	Visualization   string `bun:"-" json:"visualization"`
	Deployment      string `bun:"-" json:"deployment"`
	Infrastructure  string `bun:"-" json:"infrastructure"`
	Database        string `bun:"-" json:"database"`
	Transferability string `bun:"-" json:"transferability"`
	ContactPerson   string `bun:"-" json:"contact_person"`
	Authority       string `bun:"-" json:"authority"`
	AdditionalInfo  string `bun:"-" json:"additional_info"`
}

// AppRelationSummary is a lightweight representation of a related app.
type AppRelationSummary struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Icon string `json:"icon"`
}

// AppGroupSummary is a lightweight reference to a group an app belongs to.
type AppGroupSummary struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Icon string `json:"icon"`
}

File: services/frontend/config/apps.ts
import fs from 'fs';
import yaml from 'js-yaml';
import path from 'path';

export interface LiveDemo {
  label: string;
  url: string;
}

export interface AppLink {
  label: string;
  url: string;
}

export interface AppField {
  key: string;
  value: string;
}

export interface DeploymentVariant {
  name: string;
  description: string;
  dockerCommand?: string;
  dockerNote?: string;
  composeCommand?: string;
  composeNote?: string;
  helmCommand?: string;
  helmNote?: string;
  helmValues?: string;
}

export interface GitLabProviderSummary {
  key: string;
  type?: string;
  label: string;
  baseUrl: string;
  autoSyncEnabled?: boolean;
  syncIntervalMinutes?: number;
  defaultReadmePath?: string;
  defaultHelmValuesPath?: string;
  defaultComposeFilePath?: string;
  namespaceAllowlist?: string[];
}

export interface GitLabProviderAdminSettings {
  providerKey: string;
  providerType?: string;
  label: string;
  baseUrl: string;
  linkedAppsCount?: number;
  namespaceAllowlist: string[];
  enabled: boolean;
  autoSyncEnabled: boolean;
  syncIntervalMinutes: number;
  defaultReadmePath?: string;
  defaultHelmValuesPath?: string;
  defaultComposeFilePath?: string;
  configured: boolean;
  tokenConfigured: boolean;
}

export interface OIDCProviderSummary {
  key: string;
  label: string;
  issuer: string;
  clientId: string;
  adminGroup: string;
  insecure: boolean;
  disableLocalAuth: boolean;
  scopes: string[];
  configured: boolean;
}

export interface OIDCProviderAdminSettings {
  providerKey: string;
  label: string;
  issuer: string;
  clientId: string;
  adminGroup: string;
  enabled: boolean;
  insecure: boolean;
  disableLocalAuth: boolean;
  scopes: string[];
  configured: boolean;
  secretConfigured: boolean;
}

export interface GitLabSyncSnapshot {
  projectId?: number;
  projectName?: string;
  projectPath?: string;
  projectWebUrl?: string;
  defaultBranch?: string;
  description?: string;
  topics?: string[];
  license?: string;
  lastActivityAt?: string;
  readmePath?: string;
  readmeContent?: string;
  helmValuesPath?: string;
  helmValuesContent?: string;
  composeFilePath?: string;
  composeFileContent?: string;
  warnings?: string[];
  syncedAt?: string;
}

export interface GitLabIntegrationState {
  linked: boolean;
  availableProviders: GitLabProviderSummary[];
  providerKey?: string;
  providerType?: string;
  providerLabel?: string;
  baseUrl?: string;
  projectPath?: string;
  projectWebUrl?: string;
  branch?: string;
  readmePath?: string;
  helmValuesPath?: string;
  composeFilePath?: string;
  lastSyncStatus?: string;
  lastSyncError?: string;
  approvalRequired?: boolean;
  lastSyncedAt?: string;
  lastAppliedAt?: string;
  lastManualChangeAt?: string;
  snapshot?: GitLabSyncSnapshot;
  pendingSnapshot?: GitLabSyncSnapshot;
}

export interface GitLabSyncSummary {
  linked: boolean;
  providerKey?: string;
  providerType?: string;
  projectPath?: string;
  lastSyncStatus?: string;
  lastSyncError?: string;
  approvalRequired: boolean;
  lastSyncedAt?: string;
}

export interface AppViewerPermissions {
  canEdit: boolean;
  canDelete: boolean;
  canManageEditors: boolean;
  accessRole: 'admin' | 'owner' | 'editor' | 'viewer' | string;
}

export interface AppRelease {
  id: string;
  appId: string;
  version: string;
  releaseType: 'patch' | 'minor' | 'major' | string;
  source: string;
  title: string;
  summary: string;
  changedAreas: string[];
  changeDetails: ReleaseChangeDetail[];
  diffPreview: string;
  fingerprint: string;
  publishedAt: string;
  createdAt: string;
}

export interface ReleaseChangeDetail {
  area: string;
  field: string;
  label: string;
  language: string;
  preview: string;
  diff: string;
  beforeText: string;
  afterText: string;
}

export interface UpdatePreferences {
  userId: string;
  notifyFavoritedApps: boolean;
  notifyRecentlyViewedApps: boolean;
  notifyOwnedManagedApps: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ReleaseInboxItem {
  id: string;
  releaseId: string;
  appId: string;
  appName: string;
  appIcon?: string;
  version: string;
  releaseType: 'patch' | 'minor' | 'major' | string;
  title: string;
  summary: string;
  changedAreas: string[];
  changeDetails: ReleaseChangeDetail[];
  diffPreview: string;
  reason: string;
  publishedAt: string;
  seenAt?: string | null;
}

export interface SystemUser {
  id: string;
  username: string;
  email: string;
  role: string;
  authType?: string;
  canSubmitApps?: boolean;
  disabled?: boolean;
}

export type AppEditorUser = SystemUser;

export interface AppUserSummary {
  id: string;
  username: string;
  email: string;
}

export interface AppCatalogFilters {
  q: string;
  status: string;
  category: string;
  ownerId: string;
  hasEditors: string;
  syncStatus: string;
  featured: string;
  locked: string;
  visibility: string;
}

export interface AppConfig {
  id: string;
  name: string;
  description: string;
  categories: string[];
  liveUrl?: string; // Kept for backwards compatibility
  liveDemos?: LiveDemo[];
  repoUrl?: string;
  repositories?: AppLink[];
  customLinks?: AppLink[];
  helmRepo?: string;
  dockerRepo?: string;
  docsUrl?: string;
  icon?: string;
  techStack?: string[];
  license?: string;
  markdownContent?: string;
  /** Dynamic "Fachliche Details" key-value pairs. Schema is defined in platform settings. */
  customFields?: AppField[];
  status?: string;
  customDockerCommand?: string;
  customComposeCommand?: string;
  customHelmCommand?: string;
  customDockerNote?: string;
  customComposeNote?: string;
  customHelmNote?: string;
  customHelmValues?: string;
  hasDeploymentAssistant?: boolean;
  showDocker?: boolean;
  showCompose?: boolean;
  showHelm?: boolean;
  isLocked?: boolean;
  isReuse?: boolean;
  reuseRequirements?: string;
  ownerId?: string;
  owner?: AppUserSummary;
  editors?: AppUserSummary[];
  tags?: string[];
  collections?: string[];
  isFeatured?: boolean;
  bannerText?: string;
  bannerType?: 'info' | 'warning' | 'danger' | 'custom';
  /** Only used when bannerType is 'custom'. Expected format: #RRGGBB. */
  bannerColor?: string;
  /** Optional override for the banner heading. Defaults to the type label (Info/Warnung/Kritisch). */
  bannerTitle?: string;
  /** @deprecated Use bannerText + bannerType instead. Accepted as legacy input only. */
  knownIssue?: string;
  authority?: string;
  ratingAvg?: number;
  ratingCount?: number;
  createdAt?: string;
  updatedAt?: string;
  deploymentVariants?: DeploymentVariant[];
  version?: string;
  changelog?: string;
  // Related apps (populated by GetApp)
  relatedApps?: { id: string; name: string; icon?: string }[];
  // Groups this app belongs to
  appGroups?: { id: string; name: string; icon?: string }[];
  gitLabSync?: GitLabSyncSummary;
  viewerPermissions?: AppViewerPermissions;
  /** Opt out of the global link-probing feature for this app (e.g. links behind auth/VPN) */
  skipLinkProbe?: boolean;
  /** Backend-maintained reachability status for live endpoints. */
  linkProbeStatus?: 'ok' | 'partial' | 'down' | 'unknown';
}

export const getApps = (): AppConfig[] => {
  try {
    const configPath = process.env.APP_CONFIG_PATH || 'config.yaml';
    const filePath = path.isAbsolute(configPath) 
      ? configPath 
      : path.join(process.cwd(), configPath);
    
    if (!fs.existsSync(filePath)) {
      console.warn(`Config file not found at ${filePath}, returning empty list`);
      return [];
    }

    const fileContents = fs.readFileSync(filePath, 'utf8');
    const data = yaml.load(fileContents) as { apps: AppConfig[] };
    return data.apps || [];
  } catch (error) {
    console.error('Error loading apps from config:', error);
    return [];
  }
};

export const apps: AppConfig[] = []; // Keep as fallback or for types if needed
